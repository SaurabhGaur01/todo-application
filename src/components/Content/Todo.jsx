import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { compose } from 'redux';
import { toggleTodo } from '../../ducks/todo';
import Checkbox from '@material-ui/core/Checkbox';

/*
This component is responsible for viewing todo items -
actionToggleTodo : being used to update status to complet/incomplete in store
*/
const Todo = ({ 
    actionToggleTodo, 
    id, 
    completed, 
    text 
}) => (
    <li>
        <label style={{textDecoration: completed ? 'line-through' : 'none'}}>
            {text}
        </label>
        <Checkbox 
            onChange={() => actionToggleTodo(id)} 
            checked={completed} 
        />
    </li>
);

Todo.propTypes = {
    actionToggleTodo: PropTypes.func.isRequired,
    completed: PropTypes.bool.isRequired,
    text: PropTypes.string.isRequired,
    id: PropTypes.number.isRequired,
}

const mapDispatchAsProps = dispatch => {
    return {
        actionToggleTodo: value => dispatch(toggleTodo(value)),
    };
}

const hocChain = compose(
    connect(null, mapDispatchAsProps),
);

export { Todo as TestableTodo };

export default hocChain(Todo);
