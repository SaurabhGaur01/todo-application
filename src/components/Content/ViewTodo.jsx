import React from 'react';
import PropTypes from 'prop-types';
import Checkbox from '@material-ui/core/Checkbox';

/*
This component is responsible for displaying todo item with checkbox and label
*/
const ViewTodo = ({ 
  id, 
  onClick, 
  completed, 
  text 
}) => (
  <li key={id}>
    <Checkbox onClick={onClick} checked={completed} label={(
      <label style={{textDecoration: completed ? 'line-through' : 'none'}}>
        {text}
      </label>
    )}/>
  </li>
);

ViewTodo.propTypes = {
  id: PropTypes.number.isRequired,  
  onClick: PropTypes.func.isRequired,
  completed: PropTypes.bool.isRequired,
  text: PropTypes.string.isRequired
}

export default ViewTodo;