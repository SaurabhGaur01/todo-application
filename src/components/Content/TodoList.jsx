import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { compose } from 'redux';
import Todo from './Todo';
import { VIEW_SHOW_ACTIVE, VIEW_SHOW_COMPLETED, VIEW_SHOW_ALL } from '../../constants/view';

/*
This component is responsible for iterating and choosing selected todos to view -
*/
const TodoList = ({ 
  todos, 
  viewMode 
}) => {
  let selectedTodo = [];
  if(viewMode === VIEW_SHOW_ALL) selectedTodo = todos;
  if(viewMode === VIEW_SHOW_COMPLETED) selectedTodo = todos.filter(item => item.completed);
  if(viewMode === VIEW_SHOW_ACTIVE) selectedTodo = todos.filter(item => !item.completed);
  return (
    <ul>
      {selectedTodo.map(({ id, text, completed }) =>
          <div key={id}>
            <Todo id={id} completed={completed} text={text} />
          </div>
      )}
    </ul>
  )
};

TodoList.propTypes = {
  viewMode: PropTypes.string.isRequired,
  todos: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.number.isRequired,
    completed: PropTypes.bool.isRequired,
    text: PropTypes.string.isRequired,
  })).isRequired,
}

export const mapStateToProps = state => ({
  todos: state.todos,
  viewMode: state.viewMode,
});

const hocChain = compose(
  connect(mapStateToProps),
);

export { TodoList as TestableTodoList };

export default hocChain(TodoList);
