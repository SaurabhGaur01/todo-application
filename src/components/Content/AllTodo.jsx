import React, { useState } from "react";
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { compose } from 'redux';
import Card from '@material-ui/core/Card';
import CardContent from '@material-ui/core/CardContent';
import TextField from '@material-ui/core/TextField';
import Button from '@material-ui/core/Button';
import TodoList from './TodoList';
import { addTodo, clearTodo, massAdd } from '../../ducks/todo';

/*
This component is responsible for rendering actual body content -
inputChange : being used to get user entered text
onSubmit/ actionAddTodo : being used to update redux store & clear text area
massAction: being used to update all status on single go
clearAction: being used to clear all todos from store
*/
const AllTodo = ({ 
    actionAddTodo, 
    todos, 
    actionClearTodo, 
    actionMassAddTodo 
}) => {
    const [input, setInput] = useState(''); 

    const inputChange = event => { setInput(event.target.value); }

    const onSubmit = event => {
        event.preventDefault();
        actionAddTodo(input);
        setInput('');
    }

    const inCompleteAction = () => {
        const updatedTodos = todos.map(item => {
            var temp = Object.assign({}, item);
            temp.completed = false;
            return temp;
        });
        actionClearTodo();
        actionMassAddTodo(updatedTodos);
    };

    const completeAction = () => {
        const updatedTodos = todos.map(item => {
            var temp = Object.assign({}, item);
            temp.completed = true;
            return temp;
        });
        actionClearTodo();
        actionMassAddTodo(updatedTodos);
    };

    const clearAction = () => {
        actionClearTodo();
    };

    return (
        <div className="main-card">
            <Card className="card-container">
                <CardContent className="card-content">
                    <div className="first-section">
                        <div className="heading-label">What needs to be done?</div>
                        <form onSubmit={onSubmit}>
                            <div className="input-text">
                                <div>
                                    <TextField className="text-box" id="input" label="Input your TODOs..." onChange={inputChange} value={input} />
                                </div>
                                <span>
                                    <Button className="add-button" variant="contained" type='submit'>Add</Button>
                                </span>
                            </div>
                        </form>
                    </div>
                    <TodoList />
                </CardContent>
            </Card>
            <div className="mass-container">
                <Button className="mass-buttons" variant="contained" color="primary" onClick={completeAction}>
                    Mark all as complete
                </Button>
                &nbsp;
                <Button className="mass-buttons" variant="contained" color="primary" onClick= {inCompleteAction}>
                    Mark all as incomplete
                </Button>
                &nbsp;
                <Button className="mass-buttons" variant="contained" color="primary" onClick= {clearAction}>
                    Clear all
                </Button>
            </div>
        </div>    
    )
};

AllTodo.propTypes = {
    actionAddTodo: PropTypes.func.isRequired,
    actionClearTodo: PropTypes.func.isRequired,
    actionMassAddTodo: PropTypes.func.isRequired,
    todos: PropTypes.arrayOf(PropTypes.shape({
        id: PropTypes.number.isRequired,
        completed: PropTypes.bool.isRequired,
        text: PropTypes.string.isRequired,
    })).isRequired,
}

export const mapStateToProps = state => ({
    todos: state.todos,
});

const mapDispatchAsProps = dispatch => {
    return {
        actionAddTodo: value => dispatch(addTodo(value)),
        actionClearTodo: () => dispatch(clearTodo()),
        actionMassAddTodo: data => dispatch(massAdd(data)),
    };
}

const hocChain = compose(
    connect(mapStateToProps, mapDispatchAsProps),
);

export { AllTodo as TestableAllTodo };

export default hocChain(AllTodo);