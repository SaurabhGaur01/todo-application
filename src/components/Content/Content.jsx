import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { compose } from 'redux';
import '../TodoApp.scss';
import TabPanel from '../Tabs/TabPanel';
import TabHeader from '../Tabs/TabHeader';
import AllTodo from './AllTodo';
import { setView } from '../../ducks/view';
import { VIEW_SHOW_ACTIVE, VIEW_SHOW_COMPLETED, VIEW_SHOW_ALL } from '../../constants/view'; 

/*
This component is the entry point & responsible for rendering Tabs and content -
changeTab() : being used to switch the view based on Tab chose
actionSetView() : being used to update redux store
*/
const Content = ({ actionSetView }) => {
  const [tabValue, setTabValue] = React.useState(0);

  const changeTab = (event, newValue) => {
    setTabValue(newValue);
    if(newValue === 0) actionSetView(VIEW_SHOW_ALL);
    if(newValue === 1) actionSetView(VIEW_SHOW_ACTIVE);
    if(newValue === 2) actionSetView(VIEW_SHOW_COMPLETED);
  }

  return (
    <React.Fragment>
      <TabHeader value={tabValue} handleChange={changeTab} />
      <TabPanel value={tabValue} index={0}>
          <AllTodo />
      </TabPanel>
      <TabPanel value={tabValue} index={1}>
          <AllTodo />
      </TabPanel>
      <TabPanel value={tabValue} index={2}>
          <AllTodo />
      </TabPanel>
    </React.Fragment>
  );
}

Content.propTypes = {
  actionSetView: PropTypes.func.isRequired,
}

const mapDispatchAsProps = dispatch => {
  return {
      actionSetView: (value) => dispatch(setView(value)),
  };
}

const hocChain = compose(
  connect(null, mapDispatchAsProps),
);

export { Content as TestableContent };

export default hocChain(Content);