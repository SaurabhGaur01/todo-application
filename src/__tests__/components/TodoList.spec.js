import React from 'react';
import { shallow } from 'enzyme';
import { TestableTodoList, mapStateToProps } from '../../components/Content/TodoList';

describe('TodoList Component', () => {
    let props;

    beforeEach(() =>{
        props = {
            viewMode: 'SHOW_ALL',
            todos: [
                {
                    id: 1,
                    completed: true,
                    text: 'mockText',
                }
            ],
        }
    });

    it('should render correctly', () => {
        const renderedModule = shallow(<TestableTodoList {...props} />);
        expect(renderedModule).toMatchSnapshot();
    });

    describe('mapStateToProps', () => {
        const todosMock = [{
            id: 1,
            completed: true,
            text: 'mockText',
        }];
        
        const mockedState = {
            todos: todosMock,
        };

        it('should return the module state from the redux store', () => {
            const mockPropsFromRedux = mapStateToProps(mockedState);
            expect(mockPropsFromRedux).toEqual(
                {
                    todos: todosMock,
                }
            )
        })
    })
})