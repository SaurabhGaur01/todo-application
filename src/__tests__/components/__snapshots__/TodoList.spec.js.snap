// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoList Component should render correctly 1`] = `
<ul>
  <div
    key="1"
  >
    <Connect(Todo)
      completed={true}
      id={1}
      text="mockText"
    />
  </div>
</ul>
`;
